src/contract/CallParams.ts

```diff
@@ -0,0 +1,751 @@
+
+        return this.addParam(param);
+    }
+
+    public addBytesFixed(param: Uint8Array): this {
+        if (param.length > 32) {
+            throw new Error("`bytes<M>` type requires parameter length to be less than or equal to 32 bytes");
+        }
+
+        this.conditionallyAddType({ ty: ArgumentType.bytesfix, array: false }, param.length);
+
+        return this.addParam(param);
+    }
+
+    public addFunction(address: Uint8Array | string, func: FunctionSelector | string): this {
+        const addressParam = address instanceof Uint8Array ?
+            address :
+            Buffer.from(address as string, "hex");
+
+        const functionSelector = func instanceof FunctionSelector ?
+            (func as FunctionSelector).toProto() :
+            FunctionSelector.identifier(func as string);
+
+        if (addressParam.length !== 20) {
+            throw new Error("`function` type requires parameter `address` to be exactly 20 bytes");
+        }
+
+        this.conditionallyAddType({ ty: ArgumentType.func, array: false });
+
+        const proto = new Uint8Array(24);
+        proto.set(addressParam, 0);
+        proto.set(functionSelector, 20);
+
+        return this.addParam(proto);
+    }
+
+    public addStringArray(strArray: string[]): this {
+        this.conditionallyAddType({ ty: ArgumentType.string, array: true });
+
+        return this.addParam(strArray);
+    }
+
+    public addByteArray(byteArray: Uint8Array[]): this {
+        this.conditionallyAddType({ ty: ArgumentType.bytes, array: true });
+
+        return this.addParam(byteArray);
+    }
+
+    public addAddressArray(addresses: string[] | Uint8Array[]): this {
+        if (!Array.isArray(addresses)) {
+            throw new TypeError("string[] and Uint8Array[] are both arrays, how is this possible?");
+        }
+
+        const par: Uint8Array[] = [];
+
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        for (const [ _, a ] of addresses.entries()) {
+            if (a instanceof Uint8Array) {
+                if (a.length !== 20) {
+                    throw new Error("`address` type requires parameter to be exactly 20 bytes");
+                }
+
+                par.push(a);
+            } else {
+                const buf = Buffer.from(a as string, "hex");
+
+                if (buf.length !== 20) {
+                    throw new Error("`address` type requires parameter to be exactly 20 bytes");
+                }
+
+                par.push();
+            }
+        }
+
+        this.conditionallyAddType({ ty: ArgumentType.address, array: true });
+
+        return this.addParam(par);
+    }
+
+    // Allows user to add any parameter they wish and they type will be guessed from the
+    // type and size of the parameter passed in
+    // For more control, use the `add<Type>` methods instead.
+    public addParam(param: string | boolean | number | Uint8Array | BigNumber |
```

Comment:
I don't like this method, too easy to get wrong and too hard to debug when you do (because the function selector will change unpredictably).

If we want something like this I would force the user to specify the Solidity equivalent type instead of guessing it as they should know based on the signature of the function they're trying to invoke; however, a dynamic API like this would be difficult/verbose to duplicate in Java and our primary goal right now is parity so I'd prefer if we just dropped this one. If nothing else it deserves more thought (and not just in the context of this SDK alone) before we commit to it.

For now, I would simply add a comment somewhere to open an issue if someone needs a type that isn't provided here.